src/locate_handle.cpp

```diff
@@ -0,0 +1,210 @@
+    pcl::PointCloud<pcl::PointXYZ>::Ptr handle_cluster(new pcl::PointCloud<pcl::PointXYZ>);
+    int j = 0;
+    for (std::vector<pcl::PointIndices>::const_iterator it = cluster_indices.begin(); it != cluster_indices.end(); ++it)
+    {
+        pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_cluster(new pcl::PointCloud<pcl::PointXYZ>);
+        for (std::vector<int>::const_iterator pit = it->indices.begin(); pit != it->indices.end(); ++pit)
+            cloud_cluster->points.push_back(cloud_cropped->points[*pit]); //*
+        cloud_cluster->width = cloud_cluster->points.size();
+        cloud_cluster->height = 1;
+        cloud_cluster->is_dense = true;
+
+        Eigen::Vector4f centroid;
+        pcl::compute3DCentroid(*cloud_cluster, centroid);
+
+        // Select the cluster with centroid closest to original estimate
+        double error_x = std::abs(centroid(0) - original_handle_location.point.x);
+        double error_y = std::abs(centroid(1) - original_handle_location.point.y);
+        double error_z = std::abs(centroid(2) - original_handle_location.point.z);
+        double measured_error = error_x + error_y + error_z;
+        ROS_DEBUG_STREAM("Total error of the cluster: " << measured_error);
+        // ToDo: could also filter based on number of points to improve selection (handle ~1000 points)
+        if (measured_error < error)
+        {
+            *handle_cluster = *cloud_cluster;
+            error = measured_error;
+        }
+        ROS_DEBUG_STREAM("PointCloud representing the Cluster: " << cloud_cluster->points.size() << " data points.");
+        j++;
+    }
+
+    // check that preempt has not been requested by the client
+    if (as_->isPreemptRequested() || !ros::ok()) {
+        // set the action state to preempted
+        as_->setPreempted();
+    }
+    if (error < 1.0) {
+        // This section is called only when the cluster of the first while loop is empty (which should be never...)
+        Eigen::Vector4f min_point;
+        Eigen::Vector4f max_point;
+        pcl::getMinMax3D(*handle_cluster, min_point, max_point);
+        ROS_INFO_STREAM("Handle cluster min_point: " << min_point << ", max_point: " << max_point);
+        tue_msgs::LocateDoorHandleResult result;
+        geometry_msgs::PointStamped handle_edge_point1;
+        handle_edge_point1.header.frame_id = msg->header.frame_id;
+        handle_edge_point1.point.x = min_point(0);
+        handle_edge_point1.point.y = min_point(1);
+        handle_edge_point1.point.z = min_point(2);
+        result.handle_edge_point1 = handle_edge_point1;
+        geometry_msgs::PointStamped handle_edge_point2;
+        handle_edge_point2.header.frame_id = msg->header.frame_id;
+        handle_edge_point2.point.x = max_point(0);
+        handle_edge_point2.point.y = max_point(1);
+        handle_edge_point2.point.z = max_point(2);
+        result.handle_edge_point2 = handle_edge_point2;
```

Comment:
No need to create PointStamped and assign it to the result members after filling it. More efficient to fill the result and its members directly. (Maybe use a reference in case lines become really long.)